tools/opt/opt.cpp

```diff
@@ -0,0 +1,184 @@
+};
+
+spv_result_t BuildSpvInst(void* builder, const spv_parsed_instruction_t* inst) {
+  reinterpret_cast<ir::SpvBuilder*>(builder)->AddInstruction(inst);
+  return SPV_SUCCESS;
+};
+
+int main(int argc, char** argv) {
+  const char* in_file = nullptr;
+  const char* out_file = nullptr;
+
+  spv_target_env target_env = SPV_ENV_UNIVERSAL_1_1;
+
+  PassManager pass_manager;
+
+  for (int argi = 1; argi < argc; ++argi) {
+    const char* cur_arg = argv[argi];
+    if ('-' == cur_arg[0]) {
+      if (0 == strcmp(cur_arg, "--version")) {
+        printf("%s\n", spvSoftwareVersionDetailsString());
+        return 0;
+      } else if (0 == strcmp(cur_arg, "--help") || 0 == strcmp(cur_arg, "-h")) {
+        PrintUsage(argv[0]);
+        return 0;
+      } else if (0 == strcmp(cur_arg, "-o")) {
+        if (!out_file && argi + 1 < argc) {
+          out_file = argv[++argi];
+        } else {
+          PrintUsage(argv[0]);
+          return 1;
+        }
+      } else if (0 == strcmp(cur_arg, "--strip-debug-info")) {
+        std::unique_ptr<DebugInfoRemovalPass> pass(new DebugInfoRemovalPass);
+        pass_manager.AddPass(std::move(pass));
+      } else if (0 == strcmp(cur_arg, "--unify-types")) {
+        std::unique_ptr<TypeUnificationPass> pass(new TypeUnificationPass);
+        pass_manager.AddPass(std::move(pass));
+      } else if (0 == cur_arg[1]) {
+        // Setting a filename of "-" to indicate stdin.
+        if (!in_file) {
+          in_file = cur_arg;
+        } else {
+          fprintf(stderr, "error: More than one input file specified\n");
+          return 1;
+        }
+      } else {
+        PrintUsage(argv[0]);
+        return 1;
+      }
+    } else {
+      if (!in_file) {
+        in_file = cur_arg;
+      } else {
+        fprintf(stderr, "error: More than one input file specified\n");
+        return 1;
+      }
+    }
+  }
+
+  if (out_file == nullptr) {
+    fprintf(stderr, "error: -o required\n");
+    return 1;
+  }
+
+  std::vector<uint32_t> source;
+  const bool use_file = in_file && strcmp("-", in_file);
+  if (FILE* fp = (use_file ? fopen(in_file, "rb") : stdin)) {
+    uint32_t buf[1024];
+    while (size_t len = fread(buf, sizeof(uint32_t),
+                              sizeof(buf) / sizeof(uint32_t), fp)) {
+      source.insert(source.end(), buf, buf + len);
+    }
+    if (use_file) fclose(fp);
+  } else {
+    fprintf(stderr, "error: file does not exist '%s'\n", in_file);
+    return 1;
+  }
+
+  spv_context context = spvContextCreate(target_env);
+  spv_diagnostic diagnostic = nullptr;
+
+#if 0
+  // Let's do validation first.
+  spv_const_binary_t binary = {source.data(), source.size()};
+  spv_result_t error = spvValidate(context, &binary, &diagnostic);
+  if (error) {
+    spvDiagnosticPrint(diagnostic);
+    spvDiagnosticDestroy(diagnostic);
+    return error;
+  }
+#endif
+
+  ir::Module module;
+  ir::SpvBuilder spv_builder(&module);
+  spvBinaryParse(context, &spv_builder, source.data(), source.size(),
```

Comment:
Maybe trivial, but I think it is better to move all the parsing related code, including the two call back functions, to SpvBuilder. Just leave a BuildSpv(&module) in here.